compiler/rustc_mir_build/src/build/expr/stmt.rs

```diff
@@ -39,16 +39,29 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
                 // Generate better code for things that don't need to be
                 // dropped.
-                if lhs.ty.needs_drop(this.tcx, this.param_env) {
-                    let rhs = unpack!(block = this.as_local_operand(block, rhs));
-                    let lhs = unpack!(block = this.as_place(block, lhs));
-                    unpack!(block = this.build_drop_and_replace(block, lhs_span, lhs, rhs));
+                let needs_drop = lhs.ty.needs_drop(this.tcx, this.param_env);
+                let rhs = unpack!(block = this.as_local_rvalue(block, rhs));
+                let lhs = unpack!(block = this.as_place(block, lhs));
+
+                let source_info = if needs_drop {
+                    let span = this.tcx.with_stable_hashing_context(|hcx| {
+                        expr_span.mark_with_reason(
+                            None,
+                            DesugaringKind::Replace,
+                            this.tcx.sess.edition(),
+                            hcx,
+                        )
+                    });
+                    let source_info = this.source_info(span);
+                    unpack!(
```

Comment:
any reason you changed `build_drop_and_replace` to only include the sad path replace?

Reply:
Initially I removed all assigns and added the sad path one later. In the happy path you need the assign anyway at call site because you may not emit a replace if the type does not need a drop, so you can just add the assign regardless of whether you created a new block for a drop and replace or not.

Anyway, not a strong opinion, I'll revert to the original behavior